content/blog/2019-11-14-mempass/index.md

```diff
@@ -0,0 +1,292 @@
+
+With `i8*` pointers, all MemPass needs is the size of the memory chunk that a load or store plans to interact with. While this data is not immediately accessible, LLVM provides a handy DataLayout class. After grabbing the type of the element that the *original* pointer points to, MemPass can extract its size and pass that to our library functions.
+
+Finally, we need to actually check if we are accessing memory that is available as per the allocation table. MemPass takes the difference of the address in question with every pointer address in the allocation table, and compares it to the appropriate sizes. If the address is within the bounds of a chunk of memory, then we are fine. Otherwise, the program will emit a use after free warning, and continue to look for more vulnerabilities.
+
+> **Note**: These pointer address comparisons are somewhat sketchy, there likely may be a better approach to memory access checking that we were unable to find.
+
+## Program Termination
+
+On program termination, we need to check the allocation table for any remaining memory that has not been freed. In order to differentiate stack memory that was allocated with `alloca`, any memory allocated with `malloc` is written to a separate file, along with any pointer addresses passed to `free`. Now, MemPass just searches this file and emits any malloced addresses that were not freed. This is compiled into a final report, listing all of the memory safety vulnerabilities (among double free, use after free, and memory leak) that were detected throughout the execution of this program, along with their line numbers.
+
+## Implementation extras
+
+Another possible implementation scheme we considered was the use of a *deallocation* hashmap in addition to the allocation hashmap, to store memory that has been freed. This way, MemPass does not need add instrumentation after `alloca` instructions, it just needs to store memory addresses allocated with malloc. However, every time memory is allocated or deallocated, MemPass must check addresses in the other map to ensure there are no overlaps.
+
+Both for the proposed framework and our current one, some sort of segmentation tree implementation to store memory bounds as intervals and compare them quickly. However, the overhead of building this tree might not be worth the benefits for small programs.
+
+Another implementation idea that would have been much more effective in hindsight would be to find some way invalidate pointers after they are freed. This way, MemPass does not need to add instrumentation after every load and store (the program can exit when accesssing a specific 'invalid' pointer). 

```

Comment:
That actually seems pretty hard to accomplish! You'd need to either:

- Find all the places where that pointer exists and update it to zero or something.
- Somehow ensure that all allocations are on separate virtual memory pages, and unmap every page when the allocation it contains is deallocated.